File: RightSideViewBFS.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    '''Time Complexity: O(n)
    Space Complexity: O(n)'''
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        #BFS
        if not root:
            return None
        
        result=[]
        #insert root in the queue
        queue=deque([root])
        while len(queue)!=0:
            size=len(queue)
            for i in range(size):
                curr_node=queue.popleft()
                #If the current index and the size of the queue are equal then add
                if i ==size-1:
                    result.append(curr_node.val)
                if curr_node.left is not None:
                    queue.append(curr_node.left)
                if curr_node.right is not None:
                    queue.append(curr_node.right)
        return result
        

File: RightSideViewBFS.java
/*
* Approach:
*  1. Apply BFS like level-order traversal.
* 
*  2. Update the result list with last element of each level.
* 
* 
* Did this code successfully run on Leetcode : YES
* 
* Any problem you faced while coding this : NO
* 
* Time Complexity: O(n)
* 
* Space Complexity: O(W) == O(n)
    width of tree
* 
*/


import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class RightSideViewBFS {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();

        if (root == null)
            return result;

        Queue<TreeNode> queue = new LinkedList<>();

        queue.add(root);

        while (!queue.isEmpty()) {
            int size = queue.size();

            for (int index = 0; index < size; index++) {
                TreeNode curr = queue.poll();

                if (index == size - 1)
                    result.add(curr.val);

                if (curr.left != null)
                    queue.add(curr.left);

                if (curr.right != null)
                    queue.add(curr.right);
            }
        }

        return result;
    }
}
